flexmeasures/data/models/reporting/cost.py

```diff
@@ -0,0 +1,143 @@
+        """
+
+        production_price_sensor: Sensor = self._config.get("production_price_sensor")
+        consumption_price_sensor: Sensor = self._config.get("consumption_price_sensor")
+
+        input_sensor: Sensor = input[0]["sensor"]  # power or energy sensor
+        timezone = input_sensor.timezone
+
+        if belief_time is None:
+            belief_time = server_now()
+
+        # get prices
+        production_price = simplify_index(
+            production_price_sensor.search_beliefs(
+                event_starts_after=start,
+                event_ends_before=end,
+                beliefs_before=belief_time,
+                resolution=input_sensor.event_resolution,
+            )
+        )
+        production_price = production_price.tz_convert(timezone)
+        consumption_price = simplify_index(
+            consumption_price_sensor.search_beliefs(
+                event_starts_after=start,
+                event_ends_before=end,
+                beliefs_before=belief_time,
+                resolution=input_sensor.event_resolution,
+            )
+        )
+        consumption_price = consumption_price.tz_convert(timezone)
+
+        # get power/energy time series
+        power_energy_data = simplify_index(
+            input_sensor.search_beliefs(
+                event_starts_after=start,
+                event_ends_before=end,
+                resolution=resolution,
+                beliefs_before=belief_time,
+            )
+        )
+
+        # compute energy flow from power flow
+        if input_sensor.measures_power:
+            power_energy_data *= input_sensor.event_resolution / timedelta(hours=1)
+
+        # compute cashflow
+        # this step assumes that positive flows represent production and negative flows consumption
+        result = (
+            power_energy_data.clip(lower=0) * production_price
+            + power_energy_data.clip(upper=0) * consumption_price
+        )
+
+        results = []
+
+        for output_description in output:
+            output_sensor = output_description["sensor"]
+            _result = result.copy()
+
+            # resample result to the event_resolution of the output sensor
+            _result = _result.resample(output_sensor.event_resolution).sum()
+
+            # convert BeliefsSeries into a BeliefsDataFrame
+            _result["belief_time"] = belief_time
+            _result["cumulative_probability"] = 0.5
+            _result["source"] = self.data_source
+            _result.event_resolution = output_sensor.event_resolution
+
+            _result = _result.set_index(
+                ["belief_time", "source", "cumulative_probability"], append=True
+            )
+
+            results.append(
+                {
+                    "name": "cashflow",
+                    "column": "event_value",
+                    "sensor": output_description["sensor"],
+                    "data": _result,
+                }
+            )
+
```

Comment:
We should also validate the units of input * price = output. Not sure where that should happen, as part of the sensors are defined in the config and another part in the parameters. Maybe in the parameter schema, using the already known config, or otherwise in the compute function itself?

Reply:
I end up implementing this at the `ProfitReporter` level.